policy: expose internals of resource module policy factory to allow custom policies

wihobbs opened the pull request:
Problem: the current fluxion policy matching factory doesn't allow for customizable policies.

Add a custom option and refactor existing policies to behave similarly. Add unit testing for the string parsing convenience functions.

Update documentation for existing and new custom match policies.

Outstanding to-do:

- [x] add an option in the scheduler config to actually pass a custom string in, and probably some validation too

wihobbs commented:
And I know the testing and the code should be in separate commits... sigh (will fix)

cmoussa1 commented on resource/policies/dfu_match_policy_factory.cpp:
```diff
@@ -13,7 +13,7 @@ extern "C" {
 #endif
 }
-
+#include <iostream>
```
I'm sorry if I missed it somewhere, but are you using anything from this `#include` anywhere? Perhaps this is left over from some debugging.

cmoussa1 commented on resource/policies/dfu_match_policy_factory.cpp:
```diff
@@ -22,50 +22,98 @@ namespace resource_model {
-    resource_type_t node_rt ("node");
+std::shared_ptr<dfu_match_cb_t> create_match_cb (const std::string &policy_requested)
+{
+    std::string policy = policies.find (policy_requested)->second;
```
Should this `.find ()` call have any sort of error checking after the lookup of `policy_requested`? Maybe this is always guaranteed to find something successfully. If not, maybe you can check the contents of `policy` before moving on.

```c++
if (policy == "")
    // return an error?
    // set matcher to something default?
    // sorry that idk what the right behavior should be here
```

wihobbs replied:
Oh, actually, `known_match_policy` is called when the config is validated in `resource/modules/resource_match_opts.cpp` so it's already been validated before it gets here.

This will need additional work to support `custom`.

cmoussa1 replied:
In that case, a comment making note that it is already validated by the time it gets to that lookup might be helpful here!

cmoussa1 commented on resource/policies/dfu_match_policy_factory.cpp:
```diff
@@ -22,50 +22,98 @@ namespace resource_model {
+
+            if (option_exists ("stop_on_k_matches", policy)) {
+                ptr->set_stop_on_k_matches (parse_int_match_options ("stop_on_k_matches", policy));
+            }
```
Does this `if-else` branch need an `else` statement? Excuse my ignorance on this - I'm not sure if the addition of custom policies allows for something other than `low` or `high`

wihobbs replied:
AFAIK low and high are the only options, unless `locality` or `variation` are chosen.

cmoussa1 commented on resource/policies/base/test/matcher_policy_factory_test02.cpp:
```diff
@@ -0,0 +1,61 @@
+
+    policy_opts = policies.find ("high")->second;
+    bool fourth = Flux::resource_model::option_exists ("stop_on_k_matches", policy_opts);
+    ok (fourth == false, "policy high does not have stop_on_k_matches as substring");
```
super minor: `bool fourth` comes before `second` and `third`; not sure if this is by intention though! Feel free to ignore me

wihobbs replied:
duh. will fix

cmoussa1 commented on resource/policies/base/test/matcher_policy_factory_test02.cpp:
```diff
@@ -0,0 +1,61 @@
+    policy_opts = policies.find ("firstnodex")->second;
+    bool second = Flux::resource_model::parse_bool_match_options ("node_exclusive", policy_opts);
+    ok (second == true, "policy first uses option node_centric");
+    bool third = Flux::resource_model::option_exists ("stop_on_k_matches", policy_opts);
```
Perhaps place newlines in between each test like you do in between the first two tests?

wihobbs replied:
My convention was a newline each time the policy changed, not in between each test. However, looking at the other test, it has a newline after every `ok` so I'll stick with that convention.

jameshcorbett reviewed (commented):
Just did a quick look-through, it looks reasonable to me. I think the commits could be broken up better though (as you noted already)

trws commented on resource/policies/dfu_match_policy_factory.hpp:
```diff
@@ -24,19 +25,28 @@
-
 bool known_match_policy (const std::string &policy);
 
+const std::map<std::string, std::string> policies =
```
I'll do a deeper look over this, but first thing this should be split.  Put the declaration here with `extern` and the definition in the cpp file so this variable doesn't end up getting defined in all includers.

trws replied:
To be fair, that's what the old code did, but it shouldn't 😬

garlick commented:
PR title seems truncated?

wihobbs commented:
Well, `custom` is a new type of policy that this PR is enabling. Or will enable eventually. Maybe "resource: refactor policy options to allow a user-defined policy" is more descriptive?

Note that this is still WIP and needs some config work and testing for the custom options. Happy to incorporate any feedback on the initial draft (and thanks to those who have already reviewed!), but it's going to stay in draft state until the code can:

- [ ] accept a custom string with policy options
- [ ] reject certain strings that are requesting invalid custom options, such as `set_stop_on_k_matches=-100` and such
- [ ] test some custom options and make sure they behave as expected

wihobbs commented:
Per discussion with Tom:

- [ ] To handle custom policies, "don't validate, parse" -- parse the input on a delimiter first and then verify that each key is valid, give an error message if it isn't, possibly demote this to a warning based on a config option
- [ ] Fluxion should issue a fatal warning if a policy named is not valid -- don't fall back on first (might already be done?)
- [ ] set_stop_on_k_matches probably only takes a value of 1, however, we should test this
- [ ] we should also think about having "node_centric" have a score factor of 10000, maybe we don't want to allow users to configure this, but maybe that number doesn't cover all our bases

jameshcorbett commented on resource/policies/dfu_match_policy_factory.hpp:
```diff
@@ -24,19 +25,24 @@
 
 bool known_match_policy (const std::string &policy);
 
+bool get_match_options (const std::string match_option, std::map<std::string, bool> &options);
```
Should this be removed?

wihobbs replied:
Oh yeah, that was holdover from the first pass.

jameshcorbett reviewed (commented):
Looks pretty good! Just noticed one thing.

wihobbs commented:
I just asked @jameshcorbett to do a first pass to get me started. Note this still is in draft state until it has working sharness tests and additional reviews.

wihobbs commented:
Once the automated tests pass I'll drop draft from this PR.

trws commented:
This is starting to look really good @wihobbs!  There's one tweak I'd like to see to this so we can keep extending it.  I like the way the options are handled, this is exactly what I was hoping for.  The one tweak I'd make is to have a `policy=*` option rather than separate  `low=` and `high=`, since the base policy object types are low, high, locality and variation.  That way, we get the "you can't set both low and high at the same time" for free, and can even collapse the two low and high branches to create a matcher based on which it is then set options on the resulting object after the condition.  We can also check that policy has been set, either with the bare name or by having `policy=whatever` to allow options.  Does that sound workable?

wihobbs commented:
I think I've incorporated this feedback @trws, though slightly differently than you laid out in your comment. Instead of a `policy` key which accepts `locality`, `variation`, `high`, and `low`, I added a `policy` key that will accept `low` or `high`, and you can still set `match-policy` to `locality`, `variation`, `high`, and `low`, in which case the latter two would result in a "vanilla" `high` or `low` policy without the extra options. Does that suffice for what you laid out? It turned out to be not a huge lift.

I haven't implemented allowing `set_stop_on_k_matches` to accept an integer yet, but I'm going to go ahead and request a few reviews on this, to get feedback on the code and start getting the commits polished for approval. If you're okay with kicking that down the road to another PR (apologies, I forgot what we decided on that in our conversation yesterday), we could do that when the option itself supports integers other than 1.

cmoussa1 commented on resource/policies/dfu_match_policy_factory.cpp:
```diff
@@ -13,59 +13,168 @@ extern "C" {
+ */
+bool parse_custom_match_policy (const std::string long_string, std::map<std::string, bool> &split)
+{
+    std::vector<std::string> temp;
```
IMHO perhaps a variable name like `options` might be more descriptive of what this is going to be initialized to.

cmoussa1 commented on resource/policies/dfu_match_policy_factory.cpp:
```diff
@@ -13,59 +13,168 @@ extern "C" {
+     * abort.
+     */
+    for (int i = 0; i < temp.size (); i++) {
+        std::vector<std::string> temp_two;
```
Same suggestion here! Maybe something like `setting` or something?

cmoussa1 reviewed (commented):
Looks pretty good @wihobbs!! I've left some general feedback; I should probably defer to someone who knows the sched code better than I do to actually approve this but I think this looks pretty good to me!

wihobbs commented:
Thanks @cmoussa1! Fixed and pushed.

grondo commented on resource/policies/dfu_match_policy_factory.cpp:
```diff
@@ -13,59 +13,167 @@ extern "C" {
+         */
+        if (std::find (policy_options.begin (), policy_options.end (), settings.at (0))
+            == policy_options.end ()) {
+            std::cerr << "invalid policy option: " << settings.at (0) << std::endl;
```
Actually, you should probably avoid writing directly to stdout/err from a broker module.

If there's a way to throw an exception and catch the error message at a higher level, which would then `flux_log (LOG_ERR, ...)` the result, that might be good. Otherwise you'd have to pass down a `flux_t *` handle to this function so it can log errors.

You could also return an error via a separate parameter when the result is `false`. Some of the other Fluxion developers may have a preference here.

trws replied:
We had a chat about this yesterday after the meeting. Just above this in the call chain we're passing a std::string through by reference and appending messages to it as necessary, then returning it up to where we can handle it.  The trick is we can't always have a handle here, because this is used in both the module and in `resource_query`.  Essentially the separate parameter option.  Example of the plan on line 516 in resource_match_opts.

I have a separate issue about doing a better job of this in fluxion as a whole, but don't want it to block this PR, so plan is to pass that output string down through for now.

wihobbs replied:
Yep, that's a hole in this PR. Will fix with what we discussed. Sorry, I should have done that before marking it as "ready." I got excited [and maybe a little ahead of myself] when the testsuite _finally_ passed 😄

wihobbs replied:
One irritating thing about my implementation of this is that it takes a 3rd positional argument of `std::string &error_string` to return the error message, which, because this function is used for validation and parsing the extra options into a map, requires initializing a string in some places where we _know_ it won't be needed.

@trws's suggestion of using `std::expected<bool, std::string>` as the return type would probably be cleaner.

grondo replied:
TIL about `std::expected`!

trws replied:
Yeah, a version of that is what I want us to do.  My ideal solution would be if we could use `std::expected<RetType, std::status_code>` but `status_code` isn't standard yet, and `error_code` is... easy to misuse.  We also don't require a C++23 stdlib yet.  The `boost::outcome` with the experimental header to use `boost::outcome<RetCode, boost::status_code>` is probably best, it's very hard to misuse and much faster to compile, but it requires just a bit of setup I haven't found time to do yet.

If you're up for it @wihobbs, it's not a lot of stuff and shouldn't be complicated and would really help out other places in sched too.  That said, I'd rather that be a separate PR from this since it's pretty orthogonal.

wihobbs replied:
I opened #1330 for the task mentioned above.

trws commented on doc/man5/flux-config-sched-fluxion-resource.rst:
```diff
@@ -102,6 +102,45 @@ hinodex
+    [sched-fluxion-resource]
+
+    # system instance will use node-exclusive
+    # scheduling (with nodes of low node IDs
+    # selected first).
+    match-policy = "policy=high node_exclusive=true set_stop_on_1_matches=true"
```
```suggestion
    # scheduling first-match (with nodes of high
    # node IDs selected first).
    match-policy = "policy=high node_exclusive=true set_stop_on_1_matches=true"
```

trws reviewed (approved):
Assuming the one documentation fix, this looks really good to me. Thanks @wihobbs!

wihobbs commented:
Thanks @trws!
